### PR TITLE
Jitting Projector gives ValueError

### PR DESCRIPTION
**Pull request by AnuravModak**
This is the first draft towards #4977.

So here I have tried to solve this bug by making necessary changes in  `pennylane/ops/qubit/observables.py`, [I have not made changes in 10 files as it is showing here, would be due to the fact that i used `black -l 100 pennylane` command].

**Initialization ( `__init__` method):**

Utilized PennyLane's Wires class for wire object creation.
Replaced NumPy functions with equivalent PennyLane qml.math functions.
Checked for None state and used qml.math.isininterval for array checks.

**Eigenvalue Computation (compute_eigvals method):**

Replaced NumPy functions with PennyLane qml.math functions.
Used qml.math.where for conditional array updates.

**Justification:**

Ensured compatibility with PennyLane by replacing NumPy functions with PennyLane's qml.math equivalents.
Utilized Wires class for proper handling of wires.
Checked for None state in the initialization.

Now please review from your side and let me know if I missed anything also let me know if I need to do anything else.

I think this should most probably solve the bug you guys were looking for!!

[sc-52777]

### CODE REVIEW

**Comment by AnuravModak** [comment]
Also please let me know this

```
import jax
import pennylane as qml

import numpy as np


@qml.qnode(qml.device("default.qubit"))
def circuit(state):
    return qml.expval(qml.Projector(state, wires=0))


print(jax.jit(circuit)(np.array([0])))
```

the output for this code snippet should be `1.0` right if it works correctly??

File: pennylane/ops/qubit/observables.py [hunk 1]
```diff
@@ -435,13 +435,20 @@ class BasisStateProjector(Projector, Operation):
     # The call signature should be the same as Projector.__new__ for the positional
     # arguments, but with free key word arguments.
     def __init__(self, state, wires, id=None):
-        wires = Wires(wires)
-        state = list(qml.math.toarray(state).astype(int))
+        wires = qml.wires.Wires(wires)
 
-        if not set(state).issubset({0, 1}):
-            raise ValueError(f"Basis state must only consist of 0s and 1s; got {state}")
+        if state is None:
```
Comment (albi3ro) [inline pennylane/ops/qubit/observables.py:440]: I don't think we particularly need a check for whether or not the state is `None`.

Our philosophy on errors is:

* Is this a mistake that is hard to debug if someone makes it?
* Is it a common mistake that users have been complaining about?
* Is it something that a user might expect to be able to do, but we just haven't implemented yet?

For something like `Projector(None, wires=0)`, I think we can just let that fall over some time later in the pipeline.

Basically, we can't check everything, and checks slow things and add complexity, so only add them when they really contribute something necessary.

File: pennylane/ops/qubit/observables.py [hunk 2]
```diff
@@ -435,13 +435,20 @@ class BasisStateProjector(Projector, Operation):
...
+            raise ValueError("Basis state cannot be None")
 
-        super().__init__(state, wires=wires, id=id)
+        if not qml.math.is_abstract(state):
+            state_array = qml.math.toarray(state)
+            state_array = state_array.astype(int)
+            if not qml.math.isininterval(state_array, 0, 1).all():
+                raise ValueError(f"Basis state must only consist of 0s and 1s; got {state_array}")
+        else:
+            state_array = qml.math.asarray(state)
```
Comment (albi3ro) [inline pennylane/ops/qubit/observables.py:449]:
```suggestion
            # for cases like [jax.numpy.array(0), jax.numpy.array(1)]
            state_array = qml.math.asarray(state)
```

Might be worth adding a reminder for why we want to use `asarray` here instead of just letting `state` through unchanged.

File: pennylane/ops/qubit/observables.py [hunk 3]
```diff
@@ -435,13 +435,20 @@ class BasisStateProjector(Projector, Operation):
...
-        if not set(state).issubset({0, 1}):
-            raise ValueError(f"Basis state must only consist of 0s and 1s; got {state}")
+        if state is None:
+            raise ValueError("Basis state cannot be None")
 
-        super().__init__(state, wires=wires, id=id)
+        if not qml.math.is_abstract(state):
+            state_array = qml.math.toarray(state)
+            state_array = state_array.astype(int)
+            if not qml.math.isininterval(state_array, 0, 1).all():
```
Comment (albi3ro) [inline pennylane/ops/qubit/observables.py:446]: I can't seem to find documentation for `numpy.isininterval` or an `isininterval` in any of the ML frameworks. Mind linking some documentation to it? Would this be a custom thing we would need to add to `pennylane/math`?

**Review by albi3ro — COMMENTED** [review]
Thanks for opening this PR 🎉 

Mind running:
```
git checkout origin/master .github
git checkout origin/master doc
```
to revert linting changes in `.github` and `doc`? Maybe at some point we should start linting python files elsewhere in the repository,

**Comment by albi3ro** [comment]
> Also please let me know this
> 
> ```
> import jax
> import pennylane as qml
> 
> import numpy as np
> 
> 
> @qml.qnode(qml.device("default.qubit"))
> def circuit(state):
>     return qml.expval(qml.Projector(state, wires=0))
> 
> 
> print(jax.jit(circuit)(np.array([0])))
> ```
> 
> the output for this code snippet should be `1.0` right if it works correctly??

And yes, the answer should be `1.0` and match the behavior without jitting.

File: pennylane/ops/qubit/observables.py (hunk 3 above)
Comment (AnuravModak) [reply pennylane/ops/qubit/observables.py]: Yeah, i understand your confusion. I was working with some custom methods that i created in `pennylane.math`, then i forgot to undo the changes and rollback completely. Like deleted those custom methods in  `pennylane.math` but forgot to change it in `observables.py`, mostly because the IDE that i was using didnt mark anything. Rectified those silly typos in the recent commits. Thanks for pointing out.

File: pennylane/ops/qubit/observables.py (hunk 2 above)
Comment (AnuravModak) [reply pennylane/ops/qubit/observables.py:447]: done, please review it and let me kno.

File: pennylane/ops/qubit/observables.py (hunk 1 above)
Comment (AnuravModak) [reply pennylane/ops/qubit/observables.py]: done, please review it and let me know.

**Comment by AnuravModak** [comment]
Please let me know if everything is good with main logic in the code, then I will move fwd to make changes to pass the pipline check.

**Comment by AnuravModak** [comment]
Also please let me know why codefactor is throwing these concerns after i run the git commands below for my current branch?
Is there anything i am missing or anything to add to bypass all these?
```
git checkout origin/master .github
git checkout origin/master doc
```

Definitely didnt touch these codes as marked below:

image

File: pennylane/ops/qubit/observables.py [hunk 4]
```diff
@@ -435,13 +435,16 @@ class BasisStateProjector(Projector, Operation):
...
 
-        super().__init__(state, wires=wires, id=id)
+        if not qml.math.is_abstract(state):
+            # for cases like [jax.numpy.array(0), jax.numpy.array(1)]
+            state_array = qml.math.asarray(state)
+            state_array = state_array.astype(int)
+            if not all(x in [0, 1] for x in state_array):
+                raise ValueError(f"Basis state must only consist of 0s and 1s; got {state_array}")
+        else:
+            state_array = qml.math.asarray(state)
```
Comment (albi3ro) [inline pennylane/ops/qubit/observables.py:447]:
```suggestion
            state_array = qml.math.asarray(state)
    super().__init__(state, wires=wires, id=id)
```

looks like the `super` acidentally got removed.

**Review by albi3ro — COMMENTED** [review]
Mind please checking out master for the `.github` and `doc` folders, as those files should be left unchanged by this PR?

**Comment by AnuravModak** [comment]
> Mind please checking out master for the `.github` and `doc` folders, as though files should be left unchanged by this PR?

This time i have checkout those folders from master, but some file conflicts still persists

image

**Comment by AnuravModak** [comment]
> Mind please checking out master for the `.github` and `doc` folders, as though files should be left unchanged by this PR?

I have checked out out master for the `.github` and `doc` folders, but still `docs/readthedocs.com:xanaduai-pennylane — Read the Docs build failed!`, not sure of the root cause, only change i made is in `doc/release/changelog.md`.

**Comment by albi3ro** [comment]
I'm wondering now if you checked out your master at https://github.com/AnuravModak/pennylane/tree/master instead of the current pennylane master.  That might explain why the images got changed as well.

If you want, I can revert those files, resolve the merge conflicts, and push to your branch for you.

**Comment by Alex-Preciado** [comment]
Hi @AnuravModak 👋🏼 , thank you so much for your contribution to PennyLane 🚀!! If the last set of instructions shared by @albi3ro does not work, and given the complexity of the problem, we will we recommend opening a new PR with only the intended modifications. This will help us streamline the review process to get your contribution added before our upcoming release. Please let us know if you have any questions or need further assistance.

**Comment by AnuravModak** [comment]
> Hi @AnuravModak 👋🏼 , thank you so much for your contribution to PennyLane 🚀!! If the last set of instructions shared by @albi3ro does not work, and given the complexity of the problem, we will we recommend opening a new PR with only the intended modifications. This will help us streamline the review process to get your contribution added before our upcoming release. Please let us know if you have any questions or need further assistance.

sure I will try and do it by today 8 PM IST.
